server/api/field_note.go

```diff
@@ -0,0 +1,189 @@
+		Created:     ctx.Payload.Created,
+		UserID:      int32(ctx.Payload.UserID),
+		CategoryID:  int32(categoryId),
+	}
+
+	if ctx.Payload.Note != nil {
+		note := string(*ctx.Payload.Note)
+		fieldNote.Note = &note
+	} else {
+		fieldNote.Note = nil
+	}
+
+	if ctx.Payload.MediaID != nil {
+		mediaID := int32(*ctx.Payload.MediaID)
+		fieldNote.MediaID = &mediaID
+	} else {
+		fieldNote.MediaID = nil
+	}
+
+	if err := c.options.Database.NamedGetContext(ctx, fieldNote, "INSERT INTO fieldkit.field_note (station_id, created, user_id, category_id, note, media_id) VALUES (:station_id, :created, :user_id, :category_id, :note, :media_id) RETURNING *", fieldNote); err != nil {
+		return err
+	}
+
+	return ctx.OK(FieldNoteType(fieldNote))
+}
+
+func (c *FieldNoteController) Update(ctx *app.UpdateFieldNoteContext) error {
+	token := jwt.ContextJWT(ctx)
+	if token == nil {
+		return fmt.Errorf("JWT token is missing from context") // internal error
+	}
+
+	// TODO: use true default category ID
+	categoryId := 0
+	if ctx.Payload.CategoryID != nil {
+		categoryId = *ctx.Payload.CategoryID
+	}
+
+	fieldNote := &data.FieldNote{
+		ID:          int32(ctx.FieldNoteID),
+		StationID:   int32(ctx.Payload.StationID),
+		Created:     ctx.Payload.Created,
+		UserID:      int32(ctx.Payload.UserID),
+		CategoryID:  int32(categoryId),
+	}
+
+	if ctx.Payload.Note != nil {
+		note := string(*ctx.Payload.Note)
+		fieldNote.Note = &note
+	} else {
+		fieldNote.Note = nil
+	}
+
+	if ctx.Payload.MediaID != nil {
+		mediaID := int32(*ctx.Payload.MediaID)
+		fieldNote.MediaID = &mediaID
+	} else {
+		fieldNote.MediaID = nil
+	}
+
+	if err := c.options.Database.NamedGetContext(ctx, fieldNote, "UPDATE fieldkit.field_note SET station_id = :station_id, created = :created, user_id = :user_id, category_id = :category_id, note = :note, media_id = :media_id WHERE id = :id RETURNING *", fieldNote); err != nil {
+		return err
+	}
+
+	return ctx.OK(FieldNoteType(fieldNote))
+}
+
+func (c *FieldNoteController) Get(ctx *app.GetFieldNoteContext) error {
+	fieldNotes := []*data.FieldNoteQueryResult{}
+
+	if err := c.options.Database.SelectContext(ctx, &fieldNotes, "SELECT fn.id AS ID, fn.created, fn.user_id, fn.note, c.key AS CategoryKey, u.username, m.url AS MediaURL, m.content_type AS MediaContentType FROM fieldkit.field_note AS fn JOIN fieldkit.user u ON (u.id = fn.user_id) JOIN fieldkit.field_note_category AS c ON (c.id = fn.category_id) LEFT OUTER JOIN fieldkit.field_note_media AS m ON (m.id = fn.media_id) WHERE fn.station_id = $1", ctx.StationID); err != nil {
+		return err
+	}
+
+	return ctx.OK(FieldNotesType(fieldNotes))
+}
+
+func (c *FieldNoteController) Delete(ctx *app.DeleteFieldNoteContext) error {
+	token := jwt.ContextJWT(ctx)
+	if token == nil {
+		return fmt.Errorf("JWT token is missing from context") // internal error
+	}
+
+	if _, err := c.options.Database.ExecContext(ctx, "DELETE FROM fieldkit.field_note WHERE id = $1", ctx.FieldNoteID); err != nil {
```

Comment:
We need to get the station for this note and verify the current user has permission to do this. I've started to wrap that code in a "Permission" object. You can see an example of how it works in data.go. You create one via NewPermissions and there's a CanModifyStationBy* set of functions for checking the user can touch that station.

Reply:
The same also goes for the Add method.